### ISSUE TITLE
Can class_id==-2 in dataset?

### ISSUE THREAD

**Issue by wuhy08** [google/automl#687]
Hi @mingxingtan ,

I was looking at this line of code:

https://github.com/google/automl/blob/5ba7223a0aa74bb9d7b17b53905966c7311bdee1/efficientdet/det_model_fn.py#L250

It seems that it is trying to filter out the loss where `class_id==-2`. But when I looked at the `dataloader.py`, I could find any clues. The closest I can find is this line:

https://github.com/google/automl/blob/5ba7223a0aa74bb9d7b17b53905966c7311bdee1/efficientdet/keras/anchors.py#L214

It seems that if original `cls_targets==-1`, then it is ignored.

Does it mean the following then (`cls_targets` means the original class data without `-=1`)?

`cls_targets==-1`: IGNORED, no contribution to loss function;
`cls_targets==0`: BACKGROUND_CLASS, should detect background in forward pass, or will incur loss;
`cls_targets>0`: REAL_OBJECT, should detect correct category in forward pass, or will incur loss.

Could you confirm whether my guess is right or not?

Thank you!

**Comment by wuhy08** [comment]
I read the code a little more and now have a better understanding. 

We are using `ArgMaxMatcher` to match between `gt_boxes` and `anchors`. In the matcher's `_match` method, it has:

https://github.com/google/automl/blob/5ba7223a0aa74bb9d7b17b53905966c7311bdee1/efficientdet/object_detection/argmax_matcher.py#L144-L158

And we have `_negatives_lower_than_unmatched` always be `True` as defined in:

https://github.com/google/automl/blob/5ba7223a0aa74bb9d7b17b53905966c7311bdee1/efficientdet/keras/anchors.py#L157-L161

Also both `matched_threshold` and `unmatched_threshold` are set to `0.5`.

Then, `TargetAssigner` generate the `cls_target` here:

https://github.com/google/automl/blob/5ba7223a0aa74bb9d7b17b53905966c7311bdee1/efficientdet/object_detection/target_assigner.py#L153-L154

by using the following method:

https://github.com/google/automl/blob/5ba7223a0aa74bb9d7b17b53905966c7311bdee1/efficientdet/object_detection/target_assigner.py#L252-L255

And `self._unmatched_cls_target==0` by default:

https://github.com/google/automl/blob/5ba7223a0aa74bb9d7b17b53905966c7311bdee1/efficientdet/object_detection/target_assigner.py#L71-L74

Therefore, all the unmatched anchors (`IoU<=0.5`) should get `0` (or `-1` after `-=1`) in their `class_id`, right?

**Comment by mingxingtan** [comment]
That's right, -2 means ignore the class (sometimes they are not labelled)

**Comment by wuhy08** [comment]
Thank you!

**Comment by huangh12** [comment]
> 
> 
> That's right, -2 means ignore the class (sometimes they are not labelled)

@mingxingtan There may exists a typo in https://github.com/google/automl/blob/c64759cdbe1feac2ceb7574b91acb9a111b456fc/efficientdet/keras/anchors.py#L223 where the -2 objects are included in the num_positives.
